Add new method to handle abandoned task events

tjololo opened the pull request:
## Description
When a process task is abandoned service owners are now required to handle these in the new method ProcessTaskAbandon in ITaskProcessor.
Previous solution made it impossible for service owners to differentiate between a task being completed or abandoned.

Update to the integrations tests in app-template-dotnet is coming

## Related Issue(s)
- #25 

## Verification
- [x] **Your** code builds clean without any errors or warnings
- [x] Manual testing done (required)
- [x] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [x] All tests run green

## Documentation
- [x] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

tjololo committed:
OnAbandonProcessTask now calls out to new method ProcessTaskAbandon in ITaskProcessor

tjololo committed:
add missing EOF linebreak

RonnyB71 commented on src/Altinn.App.Core/Features/ITaskProcessor.cs:
```diff
@@ -13,4 +13,12 @@ public interface ITaskProcessor
+    /// <param name="taskId"></param>
+    /// <param name="instance"></param>
+    /// <returns></returns>
+    public Task ProcessTaskAbandon(string taskId, Instance instance);
```
This would be a breaking change if the ITaskProcessor wasn't introduced in v7 which isn't out yet. But could this have been a new interface IAbandonTask? Or do we think that is too fine grained?

tjololo replied:
Yes, would have been a breaking change if not. Smaller interfaces is a good idea, if the developer implementing both interfaces he can choose to just have one class implement both.
If we extract this to IAbandonTask we should probably rename ITaskProcessor to IEndTask

ivarne replied:
There is now default [implementations](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/default-interface-methods-versions), so adding new methods, isn’t really a big problem any more.

implementing multiple small interfaces in the same class, means you need two service registrations (one for each method used).

tjololo replied:
We opted for two interfaces. The both describe the same method.
Changed the calling code to use IEnumerable instead, this makes it possible to define one class with logic that should run in both cases just by extending both interfaces and registering two services. If custom logic for one of the process events is needed the developer can create a class implementing only one of the interfaces and register this service also

ivarne replied:
If I’m to implement two interfaces in the same class, it would be great if the methods in the interfaces would have different names. Two `ProcessEvent` methods is kind of awkward.

tjololo replied:
With the latest changes you would have to create two classes if the logic for Abandon differs from End. 
If the same code is to be executed on both you can use the same class (implementing both interfaces) register the services and the same code is executed for both events

tjololo replied:
here is an example app using the interfaces: https://altinn.studio/repos/ttd/custom-view-confirm-back/src/branch/master/App

tjololo replied:
Given it a fair amount of thought and I starting to disagree with my last change (interfaces defining the same method signature). The idea behind having two interfaces is to spare the developers having to implement methods they aren't using, seeing as most app today does not have the need for abandon hook. In my opinion two interfaces is the best way to solve that.
The initial idea behind similar method signatures was to have the possibility to create common logic, but that would come at the cost of being able to implement both interfaces in one class. Not sure if common logic for both events is a use case we would see alot or not. If we opt for different signatures the developer could implement it in one class and easily call one method from the other.
What do you think @RonnyB71 and @ivarne ?

ivarne replied:
For performance it’s a minute benefit to have small interfaces that only gets instantiated when actually required.

For discovering the available hooks, it’s better for app devs if we group multiple hooks in sensible interfaces with default implementations, so you only override what you need.

My highest priority is that we must solve similar problems in a similar way.

tjololo replied:
The main theme of v7 is smaller interfaces to solve specific tasks. So i fell this is in the theme of the other changes we have made. 
I will push a change where the method signatures will differ again so the interfaces can be implemented in one class.
My suggestion is `IProcessTaskEnd.End(taskId, instance)` and `IProcessTaskAbandon.Abandon(taskId, instance)`

tjololo replied:
The main reason for separating these is that the entire process engine is due for some rework and I don't think abandon has ha natural home in that rework.

ivarne replied:
If you think it might make sense to one day remove IProcessTaskAbanon while keeping IProcessTaskEnd, it makes a lot of sense to have them separate.

tjololo committed:
Split ITaskProcessor into two similar interfaces and support multiple implementations of both in an application

tjololo committed:
Rename process spesific interfaces to reflect this.
methods prefix changed from process to handle as interface now uses process in another context (as in BPMN process)

tjololo committed:
add linebreak at EOF

tjololo committed:
add some tests for DefaultTaskEvents.cs

tjololo committed:
Add test for new code in OnEndProcessTask

tjololo committed:
fix code smells

tjololo committed:
update abandon and end interfaces
